util/auto_skip_compressor.cc

```diff
@@ -123,9 +123,202 @@ std::unique_ptr<Compressor> AutoSkipCompressorManager::GetCompressorForSST(
+      allcompressors_.push_back(std::move(compressors_diff_levels));
+    }
+  }
+}
+
+const char* CPUIOAwareCompressor::Name() const {
+  return "CPUIOAwareCompressor";
+}
+
+Status CPUIOAwareCompressor::CompressBlock(
+    Slice uncompressed_data, std::string* compressed_output,
+    CompressionType* out_compression_type, ManagedWorkingArea* wa) {
+  // Check if the managed working area is provided or owned by this object.
+  // If not, bypass compressor logic since the working area lacks a predictor
+  if (wa == nullptr || wa->owner() != this) {
+    // We just go with the last compressor created. If Zstd is supported will be
+    // highest compression level of Zstd
+    if (allcompressors_.size() == 0) {
+      return Status::NotSupported("No compression type supported");
+    }
+    auto choosen_compression = allcompressors_index_.back();
+    size_t choosen_compression_type = choosen_compression.first;
+    size_t compression_level_ptr = choosen_compression.second;
+    return allcompressors_[choosen_compression_type][compression_level_ptr]
+        ->CompressBlock(uncompressed_data, compressed_output,
+                        out_compression_type, wa);
+  }
+  bool exploration =
+      Random::GetTLSInstance()->PercentTrue(kExplorationPercentage);
+  TEST_SYNC_POINT_CALLBACK(
+      "CPUIOAwareCompressor::CompressBlock::exploitOrExplore", &exploration);
+  auto local_wa = static_cast<CPUIOAwareWorkingArea*>(wa->get());
+  if (exploration) {
+    auto choosen_index =
+        allcompressors_index_[Random::GetTLSInstance()->Uniform(
+            static_cast<int>(allcompressors_index_.size()))];
+    size_t choosen_compression_type = choosen_index.first;
+    size_t compresion_level_ptr = choosen_index.second;
+    TEST_SYNC_POINT_CALLBACK(
+        "CPUIOAwareCompressor::CompressBlock::SelectCompressionType",
+        &choosen_compression_type);
+    TEST_SYNC_POINT_CALLBACK(
+        "CPUIOAwareCompressor::CompressBlock::SelectCompressionLevel",
+        &compresion_level_ptr);
+    return CompressBlockAndRecord(
+        choosen_compression_type, compresion_level_ptr, uncompressed_data,
+        compressed_output, out_compression_type, local_wa);
+  } else {
+    // Going with the last compressor created as the default behavior
+    auto choosen_index =
+        allcompressors_index_[allcompressors_index_.size() - 1];
+    size_t choosen_compression_type = choosen_index.first;
+    size_t compresion_level_ptr = choosen_index.second;
+    TEST_SYNC_POINT_CALLBACK(
+        "CPUIOAwareCompressor::CompressBlock::SelectCompressionType",
+        &choosen_compression_type);
+    TEST_SYNC_POINT_CALLBACK(
+        "CPUIOAwareCompressor::CompressBlock::SelectCompressionLevel",
+        &compresion_level_ptr);
+    // decide to compress
+    return CompressBlockAndRecord(
+        choosen_compression_type, compresion_level_ptr, uncompressed_data,
+        compressed_output, out_compression_type, local_wa);
+  }
+  return Status::OK();
+}
```

Comment:
It's not very clear to me that what this PR is trying to accomplish. Just the prediction, or prediction plus some exploration/exploit logic. It's not clear from the PR summary and the test too. Please draw a cleaner scope of this PR. I feel like the exploration/exploit logic is more appropriate once we introduce the goal so should be in milestone 3.2. For now let's keep this PR for prediction only.